### PR TITLE
Support any/some/all subquery.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**

### CODE REVIEW

File: expression/expressions/cmp_subquery.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+// IsStatic implements the Expression IsStatic interface.
+func (cs *CompareSubQuery) IsStatic() bool {
+	return cs.L.IsStatic() && cs.R.IsStatic()
+}
+
+// String implements the Expression String interface.
+func (cs *CompareSubQuery) String() string {
+	anyOrAll := "any"
```
Comment (siddontang) [inline expression/expressions/cmp_subquery.go]: uppercase

File: expression/expressions/cmp_subquery.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	if in == nil {
+		return nil, nil
+	}
+
+	p, err := cs.R.Plan(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	res := []interface{}{}
```
Comment (siddontang) [inline expression/expressions/cmp_subquery.go]: hold the sub select values for next check, e.g.

```
select * from t where c1 > (select c2 from t2); 
```

if not, we will run `select c2 from t2` many times.

File: parser/scanner.l [hunk 3]
```diff
@@ -459,6 +461,7 @@ sys_var		"@@"(({global}".")|({session}".")|{local}".")?{ident}
 {all}			return all
 {alter}			return alter
 {and}			return and
+{any}			return any
```
Comment (siddontang) [inline parser/scanner.l]: should here set the item like others? @shenli

Comment (shenli) [inline parser/scanner.l]: Yes. ANY is unreserved keywords, Please add it to UnReservedKeyword list and add testcase in parser_test.go.
See: https://dev.mysql.com/doc/refman/5.7/en/keywords.html

File: expression/expressions/cmp_subquery.go (hunk 1 above)
Comment (qiuyesuifeng) [inline expression/expressions/cmp_subquery.go]: Gotcha.

File: expression/expressions/cmp_subquery.go (hunk 2 above)
Comment (qiuyesuifeng) [inline expression/expressions/cmp_subquery.go]: Yes, u are right.

**Comment by qiuyesuifeng** [comment]
PTAL

**Comment by siddontang** [comment]
LGTM
Btw, IN expression uses the same way for subselect, maybe we can unify them together later.

File: expression/expressions/cmp_subquery.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	return fmt.Sprintf("%s %s %s %s", cs.L, cs.Op, anyOrAll, cs.R)
+}
+
+// Eval implements the Expression Eval interface.
+func (cs *CompareSubQuery) Eval(ctx context.Context, args map[interface{}]interface{}) (interface{}, error) {
+	if err := hasSameColumnCount(ctx, cs.L, cs.R); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	in, err := cs.L.Eval(ctx, args)
```
Comment (coocood) [inline expression/expressions/cmp_subquery.go]: I suggest rename `in` to `leftVal`.

Comment (qiuyesuifeng) [inline expression/expressions/cmp_subquery.go]: Got it, how about `lval`? `leftVal` seems a little long.

**Comment by qiuyesuifeng** [comment]
@coocood  PTAL.

File: expression/expressions/cmp_subquery.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cs.R.Value = res
+	return cs.checkResult(lv, cs.R.Value.([]interface{}))
+}
+
+func (cs *CompareSubQuery) checkAllResult(in interface{}, result []interface{}) (interface{}, error) {
```
Comment (coocood) [inline expression/expressions/cmp_subquery.go]: in -> lv

**Comment by coocood** [comment]
LGTM

**Pull request merged**